File: scripts/PyMWIR/InverseRenderer.py
import torch
from . import PyBindMWIR
from . import Scene
from . import ManyWorlds

class InverseRenderer:
    def __init__(self):
        self.inverse_renderer = PyBindMWIR.InverseRenderer()

    def Render(self, scene, many_worlds, output=None):
        if self.inverse_renderer is None:
            raise ValueError("InverseRenderer ownership has been transferred")
        if type(scene) is not Scene:
            raise TypeError("scene must be of type Scene")
        if type(many_worlds) is not ManyWorlds.ManyWorlds:    #Why not just ManyWorlds??????????????
            raise TypeError("many_worlds must be of type ManyWorlds")
        if output is not None and type(output) is not torch.Tensor:
            raise TypeError("output must be a torch.Tensor or None")

        return self.inverse_renderer.Render(scene.scene, many_worlds.many_worlds, output)

File: src/pybindmwir/inverse_renderer.cpp
#include "pybindmwir/inverse_renderer.hpp"

namespace py = pybind11;


InverseRenderer::InverseRenderer()
{
    mwir_renderer_ = std::make_unique<MWIR::InverseRenderer>();
}

torch::Tensor InverseRenderer::Render(std::shared_ptr<Scene> scene, std::shared_ptr<ManyWorlds> many_worlds, std::optional<torch::Tensor> result_tensor)
{
    if (!mwir_renderer_)
    {
        throw std::runtime_error("InverseRenderer ownership has been transferred.");
    }

    return mwir_renderer_->Render(*scene->mwir_scene_, *many_worlds->mwir_many_worlds_, result_tensor);
}


void init_inverse_renderer(py::module_ &m)
{
    py::class_<InverseRenderer, std::shared_ptr<InverseRenderer>>(m, "InverseRenderer")
        .def(py::init<>())
        .def("Render", &InverseRenderer::Render);
}